src/Build/Definition/ProjectCollection.cs

```diff
@@ -719,34 +706,18 @@ public HostServices HostServices
         {
             get
             {
-                // Avoid write lock if possible, this getter is called a lot during Project construction.
-                using (_locker.EnterDisposableUpgradeableReadLock())
+                HostServices hostServices = Volatile.Read(ref _hostServices);
+                if (hostServices != null)
                 {
-                    if (_hostServices != null)
-                    {
-                        return _hostServices;
-                    }
-
-                    using (_locker.EnterDisposableWriteLock())
-                    {
-                        return _hostServices ?? (_hostServices = new HostServices());
-                    }
+                    return hostServices;
                 }
+                hostServices = new HostServices();
+                return Interlocked.CompareExchange(ref _hostServices, hostServices, null) ?? hostServices;
```

Comment:
Constructing a new `HostServices` has no side-effects. Using `CompareExchange` to make sure that only one instance wins and becomes `_hostServices`.

Reply:
That does not seem correct, HostServices stores affinities & host services for projects - if two threads race here, we can have situations where affinities/host objects aren't set correct for the right projects.

Reply:
The logic is the same as it was before, as far as I can tell. The only difference is that the new lock-free code can create an instance of `HostServices` that ends up being thrown away.

Before:
1. Check if `_hostServices` is non-null. If so, return it.
2. Otherwise, check again under a lock, if still null create a new `HostServices`, assign it to `_hostServices` and return it.

After:
1. Check if `_hostServices` is non-null. If so, return it.
2. Create a new instance and assign it to `_hostServices` only if `_hostServices` is still null (via `CompareExchange`). If two threads race here, only one will set the field, the other one will return the result of `CompareExchange` which is the current value of the field.